internal/venice-common/src/main/java/com/linkedin/venice/helix/StringAnnotatedHelixReadOnlySchemaRepository.java

```diff
@@ -0,0 +1,150 @@
+
+import com.linkedin.venice.meta.ReadOnlySchemaRepository;
+import com.linkedin.venice.schema.SchemaEntry;
+import com.linkedin.venice.schema.rmd.RmdSchemaEntry;
+import com.linkedin.venice.schema.writecompute.DerivedSchemaEntry;
+import com.linkedin.venice.utils.Pair;
+import com.linkedin.venice.utils.concurrent.VeniceConcurrentHashMap;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+
+
+/**
+ * This class is a wrapper schema repository class, which is only used by merge conflict resolver in A/A store ingestion
+ * task. This class will annotate value schema and partial update schema so when performing Map collection merging,
+ * Map field's key can be deserialized into String type and thus is comparable for DCR purpose. Without this wrapper or
+ * user annotation, map fields will have key in UTF-8 type which is not comparable.
+ */
+public class StringAnnotatedHelixReadOnlySchemaRepository implements ReadOnlySchemaRepository {
+  private final ReadOnlySchemaRepository internalSchemaRepo;
+  private final Map<String, Map<Integer, SchemaEntry>> valueSchemaEntryMapCache = new VeniceConcurrentHashMap<>();
+  private final Map<String, Map<String, DerivedSchemaEntry>> partialUpdateSchemaEntryMapCache =
+      new VeniceConcurrentHashMap<>();
+
+  public StringAnnotatedHelixReadOnlySchemaRepository(ReadOnlySchemaRepository internalSchemaRepo) {
+    this.internalSchemaRepo = internalSchemaRepo;
+  }
+
+  @Override
+  public void refresh() {
+    internalSchemaRepo.refresh();
+  }
+
+  @Override
+  public void clear() {
+    internalSchemaRepo.clear();
+  }
+
+  @Override
+  public SchemaEntry getKeySchema(String storeName) {
+    return internalSchemaRepo.getKeySchema(storeName);
+  }
+
+  @Override
+  /**
+   * Retrieve value schema of a store and annotate its map fields. The annotation will only be done once in the repository's
+   * lifetime as the result is cached.
+   */
+  public SchemaEntry getValueSchema(String storeName, int id) {
+    if (!(valueSchemaEntryMapCache.containsKey(storeName) && valueSchemaEntryMapCache.get(storeName).containsKey(id))) {
+      SchemaEntry schemaEntry = internalSchemaRepo.getValueSchema(storeName, id);
+      SchemaEntry annotatedSchemaEntry = getAnnotatedStringMapValueSchemaEntry(schemaEntry);
+      valueSchemaEntryMapCache.putIfAbsent(storeName, new VeniceConcurrentHashMap<>());
+      valueSchemaEntryMapCache.get(storeName).put(id, annotatedSchemaEntry);
+    }
+    return valueSchemaEntryMapCache.get(storeName).get(id);
+  }
+
+  @Override
+  public boolean hasValueSchema(String storeName, int id) {
+    return internalSchemaRepo.hasValueSchema(storeName, id);
+  }
+
+  @Override
+  public int getValueSchemaId(String storeName, String valueSchemaStr) {
+    return internalSchemaRepo.getValueSchemaId(storeName, valueSchemaStr);
+  }
+
+  @Override
+  public Collection<SchemaEntry> getValueSchemas(String storeName) {
```

Comment:
Will DCR use this method? If yes, the returned schemas won't be annotated.
If not, let us throw an exception here to make sure DCR won't use. this code path in the future, and if there is a requirement to use this function, we need to annotate the returned schema collection.

Reply:
Hi Gaojie, I think according to the discussion before, DCR schema does not have map, so we can only look at user schema.

Reply:
Well, this method is returning user's value schemas, no?

Reply:
Oh I see what you mean. This API I don't think is used by SIT at all. This is used by controller and related sub-components
I think only used APIs in SIT are getValueSchema and getDerivedSchema (well it is partial update schema, we should rename this holistically in another PR).

Reply:
If this one is not being used, shall we throw exception instead of returning un-annotated value schemas to avoid future mistakes since the user of this class may assume all the returned schemas are annotated.